### ISSUE TITLE
Proposal: Simplified configuration

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#3626]
**Problem:** We're getting more complaints about the severity code (0, 1, 2) in configuration files. While I personally don't see this as a big problem, and people do seem to get used to it, it's still confusing to new users. 

**Proposal:** We augment configuration with some additional options. In addition to `rules`, we introduce `warnings` and `errors`, so you can place a rule directly into the group you want it in. Usage would entail:

1. A rule ID can be set to `true` to enable in that severity. If the rule has options, then providing the options (array, string, object) is equivalent to `true`.
1. A rule ID can be set to `false` to disable it in that severity.
1. Priority goes in order of: `errors`, `warnings`, and `rules`. That means the setting in `errors` overrides everything else, and `warnings` overrides `rules`.

What it would look like:

```yaml
rules:
    indent: [2, 4]
warnings:
    indent: false        # actually, turn off indent
    semi: "always"
errors:
    quotes: "double"
    semi: "never"      # actually, we should never use semicolons
```

We would also then have to introduce some new inline comments, like:

```js
/*eslint-errors indent: 4, semi: "never"*/
/*eslint-warnings indent: 4, semi: "never"*/
```

**Implementation:** In order to keep things simple, we could just augment configurations as they come, pulling out `errors` and `warnings`, converting their options to an array and prepending the correct severity code. Basically, preprocess new-style configs into old-style configs and then let the old-style configs flow through the rest of ESLint.

**Points to investigate:**

1. If we have any rules whose option is simply a boolean value, as that would clash with this proposal. If yes, then we'd need to update that rule(s) so it doesn't accept a boolean value.
1. Is it confusing to be able to accidentally put the same rule in more than one section? Do we need some sort of validation to prevent that?
1. Does the preprocessing of configs have a significant performance impact?

**Comment by ilyavolodin** [comment]
This is in regards to point 2.
```yaml
rules:
    indent: [0,4]
warnings:
    indent: true
errors:
    indent: true
```

What's the expected behavior of the example above?
Sorry, I'm not sure I like this proposal. If I remember correctly your argument to proposal to replace [0,1,2] with string values ['off', 'warning', 'error'] was that you don't want arbitrary strings in configuration. How is having arbitrary strings as section names any different?
I agree that 0,1,2 might be confusing to new users. I'm just not sure I see a better solution.

**Comment by nzakas** [comment]
The expected behavior is that `indent` is an error (as mentioned, `errors` overrides `warnings`, which overrides `rules`.

> If I remember correctly your argument to proposal to replace [0,1,2] with string values ['off', 'warning', 'error'] was that you don't want arbitrary strings in configuration. How is having arbitrary strings as section names any different?

Not quite correct. My argument was that I don't want arbitrary *repeated* strings, because that increases the frequency of typos. Arbitrary one-time strings are a different thing. 

> I agree that 0,1,2 might be confusing to new users. I'm just not sure I see a better solution.

That's why I'm proposing something. You should propose something, too. I'm not saying we need to change, but I want to at least explore the possibility of a backwards-compatible change and this is the best I could come up with.

Maybe everyone hates it, that's fine we don't have to do it. Maybe it spurs another idea that is better, that's great, we can explore it. 

I just want to make sure we're doing due diligence on this problem and not just throwing our hands up and saying, "that's the way it's always been, so that's the way it will always be."

**Comment by IanVS** [comment]
> Is it confusing to be able to accidentally put the same rule in more than one section?

I would have to say "yes" to that question.  Even in your short example, I felt slightly overwhelmed trying to reason about what the final result would be.  I personally like how clean the config is currently.  

What if, instead of allowing `rules`, `warnings`, and `errors` all in the same file, config files could use one of two styles:

1. `rules` only (current style)
1. `errors` & `warnings` (cannot include `rules`)
    * Rules can only appear in one place, either `errors` or `warnings`, but not both.

This would mean that whenever you see a rule in the config file, you could immediately know exactly what the severity is, without needing to think about priority and overrides.  I like the idea of adding another way for people to set up their configuration, as long as there are well-defined guard-rails that keep people from making crazy configs that are difficult to reason about.

Those are my :two: cents.

**Comment by ilyavolodin** [comment]
@IanVS That wouldn't work if you extend an existing config and want to turn one of the rules off.

P.S. I don't mean to be negative and turn down all of the ideas without suggesting alternatives, I just can't come up with a good alternative.

**Comment by BYK** [comment]
I liked @IanVS' suggestion a lot but also haven't thought about extending other rules. May be we can have another section, `disabled` to allow this?

**Comment by lo1tuma** [comment]
I think multiple ways of doing the same thing would increase the confusion about eslint configs.

**Comment by btmills** [comment]
Regarding @IanVS's suggestion

> What if, instead of allowing rules, warnings, and errors all in the same file, config files could use one of two styles

and the issue brought up by @ilyavolodin:

> That wouldn't work if you extend an existing config and want to turn one of the rules off.

It would work if reading a config file accepts only one form (`rules` or `errors/warnings`) with that validation occurring before merging configs. After reading we'd convert to the internal 0/1/2 representation and then merge. That way we wouldn't even have to touch the merging logic.

**Comment by ilyavolodin** [comment]
@btmills Sorry, the issue that I was trying to bring up, is that if you are using `errors/warnings` style in your config, and also extending some other config, you do not have a way to disable a rule from shared config, since `errors/warnings` don't provide a mechanism for setting a rule to off.

**Comment by btmills** [comment]
@ilyavolodin Ah! Now what @BYK was saying about `disabled` makes so much more sense.

**Comment by IanVS** [comment]
So, the new style might include `errors`, `warnings`, and `disabled` (or maybe `off`)?  That seems reasonable to me.

**Comment by xjamundx** [comment]
What about a simpler `true` = `2` approach?

**Comment by ilyavolodin** [comment]
While I don't think `true` = `2` approach is ideal, I think I like `true` = `2`, `false` = `0` approach better. I think it would be clearer, and doesn't require a whole new style.

**Comment by nzakas** [comment]
I have an allergy towards values that can be multiple different types, because I don't think it's logical to assume `true` is 2 instead of 1 when `false` is 0.

Regarding exclusivity of `rules` vs. `errors/warnings`: not sure about that. I don't know that it actually solves any problems other than some squishy clean feeling. I don't think having more than one way to do something is necessarily bad in this case: we need to deprecate one way and promote the new way. This is generally how deprecation works.

Not a fan of `disabled`, because what would the value mean? `true` means disable and `false` means what? I'd rather just have rules set to false in either `errors` or `warnings` indicate "disabled".

However, it seems like there's not a lot of support for this proposal, so I'm going to consider it benched and leave this open for counter-proposals.

**Comment by IanVS** [comment]
> Regarding exclusivity of  rules  vs.  errors/warnings : not sure about that.  I don't know that it actually solves any problems other than some squishy clean feeling. 

As long as rules cannot exist in more than one location in the config file, and rules specified anywhere in the file override rules specified in extended configurations, I think you're right that there's not a big benefit.

In my mind, for tri-state configuration, it sure is hard to beat 0, 1, 2.  Maybe the solution to confusion is better training/docs?

**Comment by gyandeeps** [comment]
I see this problem as a change problem. How change is not very nice. Like if you start using 0, 1 and 2 then you kinda get comfortable with it after a while. I bet if we come up with a new system then that change would again be a bottle neck for people too. This is my thought.

In India I used to drive on the left side of the road, but when I came to US then I had to drive on right side. Took a little bit time but now it doesnt feel awkward. Just a thought.

**Comment by nzakas** [comment]
I mostly agree. The complaint keeps coming up, though, so I want to be sure we really think there's no other options before dismissing.

**Comment by BYK** [comment]
I feel like `errors/warnings/disabled` syntax is a nice way forward.

**Comment by SpadeAceman** [comment]
> My argument was that I don't want arbitrary *repeated* strings, because that increases the frequency of typos. Arbitrary one-time strings are a different thing.

How about single-character codes "e" (error), "w" (warning) and "d" (disabled)? Or would single-character strings like this still be considered arbitrary repeated strings?

I wonder, though, if the confusion doesn't stem from the 0/1/2 severity codes often being side-by-side with the other options for the rule itself. Rules like those below illustrate the problem -- it's difficult to see, at a glance, which numbers mean what for these rules:
```
// ...
,"indent": [2, 2]
,"max-len": [2, 80, 2, {
  "ignoreComments": true
}]
// ...
```
If the arguments were more clearly separated from the severity codes, it might help:
```
// ...
,"indent": [2, {
  "args": 2
}]
,"max-len": [2, {
  "args": [80, 2]
  ,"ignoreComments": true
}]
// ...
```
Then again, that's more visually cluttered when it's in JSON format. Perhaps someone else has a better idea of how to separate the severity codes from the rule options?

(FWIW, I'm also not a fan of the original proposal, as segregating rules into different sections based on severity code wouldn't allow me to keep related rules next to one another. Plus, in real-world config files, a rule's severity code would no longer be immediately obvious -- you'd have to scroll up a ways to find out which section it's in.)

**Comment by platinumazure** [comment]
What about just being more explicit? Unfortunately this means some repeated strings and also a breaking change, but...

Example (in JSON-ish):

```js
"rules": {
    "quotes": {
        "severity": "error", // or 2, or...
        "options": ["double"]
    }
}
```

My guess is this will be panned as having all of the flaws of the other proposals and few of the benefits, but at least you can't argue against its readability.

**Comment by nzakas** [comment]
That goes back to my problem with repeated string values. The more words you force people to type, the more typos they make.

**Comment by nzakas** [comment]
Okay, I think I've found some religion around this. People don't like 0, 1, 2 and would rather there be words. Previously, this was a problem because a) people can misspell things easily and b) we didn't used to validate configs, so invalid settings meant "off". Now that we do validate configs for correct values, that should at least reduce the pain associated with typos. 

So, I think we can probably support "off", "warning", "error" in addition to 0, 1, 2 and by just converting the former into the latter when we process the config. So internally, we'd still just be using 0, 1, and 2, but externally those could be represented by "off", "warning", and "error".

Thoughts?

**Comment by gyandeeps** [comment]
Sounds good to me.
Case sensitive?

**Comment by hzoo** [comment]
Allowing for any case sounds fine. Could just be `"option.toLowerCase() === "off"` then it's 0, etc? Sounds good to allow users to pick what they feel is comfortable while the internals don't necessarily have to change

**Comment by nzakas** [comment]
We can do case insensitive to make it easy. Final list:

* "off"
* "warn"
* "error"

Maybe in the future we can also have "fix", but not right now. :)

**Comment by ilyavolodin** [comment]
Sounds good to me.

**Comment by xjamundx** [comment]
I like this a lot :)

**Comment by xjamundx** [comment]
Would there be any sort of complaining for invalid values?

**Comment by ilyavolodin** [comment]
Schema validation will complain automatically. We don't even need to implement anything for that.

**Comment by xjamundx** [comment]
@ilyavolodin :+1:
